### PR TITLE
Fix assignment chains

### PR DESCRIPTION
**Pull request by seanyoung**
Fix assigning to both a storage variable and non-storage in assignment chain.

Fixes https://github.com/xermicus/fuzzy-sol/issues/142

Signed-off-by: Sean Young <sean@mess.org>

### CODE REVIEW

File: src/sema/expression.rs [hunk 1]
```diff
@@ -4260,11 +4260,13 @@ fn assign_single(
...
             }
 
+            let ty = ty.deref_any();
+
             Ok(Expression::Assign(
                 *loc,
                 ty.clone(),
                 Box::new(var.clone()),
-                Box::new(val.cast(&right.loc(), ty.deref_any(), true, ns, diagnostics)?),
+                Box::new(val.cast(&right.loc(), ty, true, ns, diagnostics)?),
```
Comment (xermicus) [inline src/sema/expression.rs:4269]: Out of curiosity, why is this changed?

**Review by xermicus — APPROVED** [review]
LGTM

File: src/sema/expression.rs (hunk 1 above)
Comment (seanyoung) [reply src/sema/expression.rs:4269]: The problem is that the assignment expression `a = c` should have the same type as the value being assigned to `a`. I first changed it to:

```
Ok(Expression::Assign(
                *loc,
                ty.deref_any().clone(),
                Box::new(var.clone()),
                Box::new(val.cast(&right.loc(), ty.deref_any(), true, ns, diagnostics)?),
                Box::new(val.cast(&right.loc(), ty, true, ns, diagnostics)?),
```
But this means `deref_any()` is being called twice, so avoided that with a let binding.

Comment (seanyoung) [reply src/sema/expression.rs:4269]: In other words, this changes the second field of Expression::Assign to the right value.

Comment (xermicus) [reply src/sema/expression.rs:4269]: Ah yes I get it now :facepalm: thanks

**Review by LucasSte — APPROVED** [review]
